File: vi/codes/swift/chapter_computational_complexity/worst_best_time_complexity.swift
/**
 * File: worst_best_time_complexity.swift
 * Created Time: 2022-12-26
 * Author: nuomi1 (nuomi1@qq.com)
 */

/* Tạo một mảng gồm các phần tử { 1, 2, ..., n } với thứ tự xáo trộn */
func randomNumbers(n: Int) -> [Int] {
    // Tạo mảng nums = { 1, 2, 3, ..., n }
    var nums = Array(1 ... n)
    // Xáo trộn ngẫu nhiên các phần tử của mảng
    nums.shuffle()
    return nums
}

/* Tìm chỉ số của số 1 trong mảng nums */
func findOne(nums: [Int]) -> Int {
    for i in nums.indices {
        // Khi phần tử 1 nằm ở đầu mảng, đạt độ phức tạp thời gian tốt nhất O(1)
        // Khi phần tử 1 nằm ở cuối mảng, đạt độ phức tạp thời gian tệ nhất O(n)
        if nums[i] == 1 {
            return i
        }
    }
    return -1
}

@main
enum WorstBestTimeComplexity {
    /* Mã điều khiển */
    static func main() {
        for _ in 0 ..< 10 {
            let n = 100
            let nums = randomNumbers(n: n)
            let index = findOne(nums: nums)
            print("Array [ 1, 2, ..., n ] after shuffling = \(nums)")
            print("Index of number 1 is \(index)")
        }
    }
}

File: vi/codes/dart/chapter_computational_complexity/worst_best_time_complexity.dart
/**
 * File: worst_best_time_complexity.dart
 * Created Time: 2023-02-12
 * Author: Jefferson (JeffersonHuang77@gmail.com)
 */

/* Tạo mảng có các phần tử { 1, 2, ..., n }, thứ tự bị xáo trộn */
List<int> randomNumbers(int n) {
  final nums = List.filled(n, 0);
  // Tạo mảng nums = { 1, 2, 3, ..., n }
  for (var i = 0; i < n; i++) {
    nums[i] = i + 1;
  }
  // Xáo trộn ngẫu nhiên các phần tử của mảng
  nums.shuffle();

  return nums;
}

/* Tìm chỉ số của số 1 trong mảng nums */
int findOne(List<int> nums) {
  for (var i = 0; i < nums.length; i++) {
    // Khi phần tử 1 nằm ở đầu mảng, đạt được độ phức tạp thời gian tốt nhất O(1)
    // Khi phần tử 1 nằm ở cuối mảng, đạt được độ phức tạp thời gian xấu nhất O(n)
    if (nums[i] == 1) return i;
  }

  return -1;
}

/* Driver Code */
void main() {
  for (var i = 0; i < 10; i++) {
    int n = 100;
    final nums = randomNumbers(n);
    int index = findOne(nums);
    print('\nArray [ 1, 2, ..., n ] after shuffling = $nums');
    print('Index of number 1 is + $index');
  }
}
